changed_from_main.py

```diff
@@ -0,0 +1,89 @@
+"""
+This script determines whether one of the projects in the repo or any of its
+dependencies are different from the version on the `main` branch.
+
+This is useful for running tests only on projects that have changed.
+
+This script should depend only on Python 3.6+ standard libraries.
+"""
+import argparse
+import re
+import os
+import subprocess
+from typing import Dict, Sequence
+import sys
+
+
+DIRNAME = os.path.dirname(os.path.abspath(__file__))
+DEPENDENCIES_DOTFILE = os.path.join(DIRNAME, "dependencies.dot")
+
+DEFINITION_PATTERN = re.compile(r"\s*([a-zA-Z0-9]+) \[.*label=\"(.*)\".*\]")
+DEPENDENCY_PATTERN = re.compile(r"\s*([a-zA-Z0-9]+) -> ([a-zA-Z0-9]+)")
+
+
+def get_dependencies() -> Dict[str, Sequence[str]]:
+    name_to_subdir = {}
+    subdir_dependencies = {}
+    with open(DEPENDENCIES_DOTFILE, "r") as f:
+        dotfile_text = f.read()
+    for groups in DEFINITION_PATTERN.findall(dotfile_text):
+        name_to_subdir[groups[0]] = groups[1]
+        subdir_dependencies[groups[1]] = []
+    for groups in DEPENDENCY_PATTERN.findall(dotfile_text):
+        name = groups[0]
+        dependency_name = groups[1]
+        if name in name_to_subdir:
+            subdir = name_to_subdir[name]
+            dependency_subdir = name_to_subdir[dependency_name]
+            subdir_dependencies[subdir].append(dependency_subdir)
+    return subdir_dependencies
+
+SUBDIR_DEPENDENCIES = get_dependencies()
```

Comment:
the reason this is globally defined is because it is used in the argparse, right? 

would it nevertheless be cleaner if we call `get_dependencies` first and pass this into parse args? It feels slightly unnatural to have this be defined globally.

Reply:
Good idea. Updated.